Energyid integration

Molier opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Introducing the [EnergyID](https://www.energyid.eu/) integration! EnergyID is an online energy monitoring platform. This integration allows you send sensors from within Home Assistant (eg. energy meters, solar panels, batteries, ...) to your EnergyID-record, using their [webhooks](https://app.energyid.eu/integrations/webhookin).

* Some reasons why you would want to upload data (specifically energy measurements) to EnergyID:

    * To store your data online, as a backup
    * To use the EnergyID web-app and/or mobile app. We have many analyses, dashboards, reports you might want to use.
    * To participate in energy communities via EnergyID (energy sharing, virtual power plants etc.)
    * Benchmarking of energy data with other users on EnergyID
    * Data portability (people want to get data into EnergyID that they can only get from within their local network, hence they want to leverage Home Assistant)


As seen before in [some threads](https://community.home-assistant.io/t/energyid-support/135851/2) and git repos ([1](https://github.com/tijsverkoyen/HomeAssistant-EnergyID) , 2) there was some demand for an official integration for data -> EID and not just the other way around.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/37472
- Link to developer documentation pull request: 
- Link to frontend pull request: https://github.com/home-assistant/brands/pull/6539

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

JrtPec committed:
energyid integration

JrtPec committed:
energyid integration

JrtPec committed:
100% test coverage

JrtPec committed:
100% test coverage

JrtPec committed:
move config names to const

JrtPec committed:
move config names to const

JrtPec committed:
add energyid to brands

JrtPec committed:
add energyid to brands

JrtPec committed:
simplify tests

JrtPec committed:
simplify tests

JrtPec committed:
group api errors

JrtPec committed:
group api errors

JrtPec committed:
uncapitalize strings

JrtPec committed:
uncapitalize strings

JrtPec committed:
remove unused entries in manifest

JrtPec committed:
remove unused entries in manifest

JrtPec committed:
remove options flow

JrtPec committed:
remove options flow

Molier committed:
chore: added init for the tests

Molier committed:
chore: added init for the tests

Molier committed:
chore: add CODEOWNERS entry for energyid tests

Molier committed:
chore: add CODEOWNERS entry for energyid tests

Molier committed:
remove: delete energyid brand configuration file

Molier committed:
remove: delete energyid brand configuration file

Molier committed:
refactor: update type hints to use union syntax for optional parameters

Molier committed:
refactor: update type hints to use union syntax for optional parameters

Molier committed:
refactor: update return type for async_step_user to use ConfigFlowResult

Molier committed:
refactor: update return type for async_step_user to use ConfigFlowResult

Molier committed:
refactor: update exception handling in test for async_setup_entry

Molier committed:
refactor: update exception handling in test for async_setup_entry

Molier committed:
refactor: enhance documentation and error handling in EnergyID integration
feat: 100 code cov

Molier committed:
refactor: enhance documentation and error handling in EnergyID integration
feat: 100 code cov

Molier committed:
refactor: update webhook URL initialization in ConfigFlow to use an empty string

Molier committed:
refactor: update webhook URL initialization in ConfigFlow to use an empty string

Molier committed:
refactor: update EnergyID integration with new webhook requirements
feat: add quality scale documentation

Molier committed:
refactor: update EnergyID integration with new webhook requirements
feat: add quality scale documentation

Molier committed:
refactor: update iot_class to cloud_polling

Molier committed:
refactor: update iot_class to cloud_polling

Molier committed:
refactor: add integration_type to EnergyID manifest

Molier committed:
refactor: add integration_type to EnergyID manifest

Molier committed:
Update quality scale status and add common fixtures for EnergyID tests

Molier committed:
Update quality scale status and add common fixtures for EnergyID tests

Molier committed:
Fix capitalization in EnergyID webhook strings

Molier committed:
Fix capitalization in EnergyID webhook strings

Molier committed:
refactor: add runtime_data parameter to MockEnergyIDConfigEntry and update tests for 100 percent coverage

Molier committed:
refactor: add runtime_data parameter to MockEnergyIDConfigEntry and update tests for 100 percent coverage

Molier committed:
refactor: quality scale update and strings clarification for consistency

Molier committed:
refactor: quality scale update and strings clarification for consistency

Molier committed:
refactor: enhance webhook connection handling and implement retry logic for uploads

Molier committed:
refactor: enhance webhook connection handling and implement retry logic for uploads

Molier committed:
refactor: add connection recheck reconnect features and keep 100 test cov

Molier committed:
refactor: add connection recheck reconnect features and keep 100 test cov

zweckj reviewed (changes requested):
Hi, thanks for your contribution 👋

So

- why would I want to sent something from my HA instance somewhere? Can you please explain a bit more about the use-case and what data/entities would be supported?
- The library code is not properly linked in mypy and doesn't use tagged releases which are pushed using CI (which we required)
- If we are building an integration that doesn't have any entities but only provides services we should implement, well, proper services (in services.py with a services.yaml, which are added in async_setup)

JrtPec commented:
@zweckj

- Some reasons why you would want to upload data (specifically energy measurements) to EnergyID:
    - To store your data online, as a backup
    - To use the EnergyID web-app and/or mobile app. We have many analyses, dashboards, reports you might want to use.
    - To participate in energy communities via EnergyID (energy sharing, virtual power plants etc.)
    - Benchmarking of energy data with other users on EnergyID
    - Data portability (people want to get data into EnergyID that they can only get from within their local network, hence they want to leverage Home Assistant)
- By library code, you mean [energyid-webhooks](https://pypi.org/project/energyid-webhooks/)? We can add type hinting, tagged releases, CI... shouldn't be any issue but please confirm this is what you mean.
- Can you clarify about services? (are they now called Actions?). You would like this integration implemented as such?

JrtPec reviewed (approved):
I have tested the integration and could successfully upload sensor data to EnergyID!

zweckj commented:
> @zweckj
> 
> * Some reasons why you would want to upload data (specifically energy measurements) to EnergyID:

check ✅

> * By library code, you mean [energyid-webhooks](https://pypi.org/project/energyid-webhooks/)? We can add type hinting, tagged releases, CI... shouldn't be any issue but please confirm this is what you mean.

Yes exactly. Type hinting is nice to have and only a requirement for `platinum` quality. Tagged releases and CI is required for every new integration.

> * Can you clarify about services? (are they now called Actions?). You would like this integration implemented as such?

Exactly what I was thinking about, but now that I understand the use-case a bit better not fitting. So you'd basically select one or multiple entities and the integration would continuously push the updates to the cloud right? In that case a service (action) wouldn't be fitting as those are one time calls triggered by the user

Molier committed:
refactor: update energyid quality scale to silver and bump webhooks requirement to 0.0.8

Molier committed:
refactor: update energyid quality scale to silver and bump webhooks requirement to 0.0.8

Molier commented:
implemented tagged releases and typing in linked library

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,105 @@
+        # Show the form
+        data_schema = vol.Schema(
+            {
+                vol.Required(CONF_WEBHOOK_URL): str,
+                vol.Required(CONF_ENTITY_ID): vol.In(hass_entity_ids(self.hass)),
+                vol.Required(CONF_METRIC): vol.In(sorted(meter_catalog.all_metrics)),
+                vol.Required(CONF_METRIC_KIND): vol.In(ENERGYID_METRIC_KINDS),
+                vol.Required(CONF_UNIT): vol.In(sorted(meter_catalog.all_units)),
```
ok let's take some design decisions before we dive into the actual review. We probably shouldn't create an entry per mapping that's just noisy. How many `METRIC_KINDS` does your platform offer? How many of those mappings would you realistically expect for an average users?

Molier replied:
* Each config entry represents one straightforward data flow from HA to EnergyID
* The metric kinds (cumulative, total, delta, gauge) are just attributes of that flow, *not* separate mappings
* Users only need exactly one config entry per entity they want to monitor, a typical user will have a few of those. (fe. energy offtake, injection,solar production,gas usage etc)

We are also working on a new webhook implementation that will significantly simplify this process. The new system will use predefined metric types for common energy measurements (electricity, gas, solar, etc.), eliminating the need for users to manually specify metric kinds and units in many cases.

zweckj replied:
That's exactly what I meant: One config entry per metric doesn't make sense, but to decide on the proper way forward I would like to know how many montiored entities per user are likely to be expected

JrtPec replied:
I would on average expect 1 to 5 entities to be monitored. In some use cases this will be more, i.e. users who also do a lot of submetering for heat pumps, ev chargers, HVAC...

I do agree that, especially with a larger numbers of entities, multiple config entries may not be as elegant as a single one. But it is a simple and transparent implementation: a single entry tracks 1 entity for updates and posts changes - within the rate limits - to EnergyID.

You propose to have a single config entry that tracks multiple entities? I recon you could store a list of entities and listen to state changes for them all? The config flow would be more complex though...

Molier replied:
Thanks for the feedback! I'll implement the config subentries approach as you suggested. This will provide a much better user experience and align with Home Assistant's design direction aswell.
I'll restructure the integration to:

* Create one main config entry for the webhook connection
* Use config subentries for each entity mapping

Also will do some improvements to the webhooks, 
* Implement aggregation for the webhook calls by collecting entity data at fixed intervals rather than sending updates on every state change
* Group entity updates by timestamp for more efficient data transmission

This approach will give users a cleaner interface where they add one EnergyID connection and then manage their entity mappings as subentries under it.
I'll be going back to draft status briefly to implement these changes. The updated design will be more scalable and provide a better foundation for future enhancements.

Molier committed:
feat: MVP working with new webhooks, sensor can be added and sync works with EID

Molier committed:
feat: MVP working with new webhooks, sensor can be added and sync works with EID

Molier committed:
feat: Integration now allows for adding, updating and removing mappings. also 1 sensor added to track your mappings

Molier committed:
feat: Integration now allows for adding, updating and removing mappings. also 1 sensor added to track your mappings

Molier committed:
chore: reloading works properly

Molier committed:
chore: reloading works properly

Molier committed:
feat: complete reworking of the integration to work with new webhook implementation

Molier committed:
feat: complete reworking of the integration to work with new webhook implementation

Molier committed:
Update pyproject.toml

Molier committed:
Update pyproject.toml

Molier commented on pyproject.toml:
```diff
@@ -124,6 +124,7 @@ dependencies = [
   "yarl==1.20.0",
   "webrtc-models==0.3.0",
   "zeroconf==0.147.0",
+  "energyid-webhooks>=0.0.14",
```
Has been removed, but where do i also need to add the dep to make sure the generate works? Since the run failed, i made another mistake?

MartinHjelmare replied:
Run the script `python3 -m script.gen_requirements_all` and commit the changes.

Molier committed:
undid all changes to pyproject.toml

Molier committed:
undid all changes to pyproject.toml

Molier committed:
Remove energyid-webhooks dependency from requirements.txt

Molier committed:
Remove energyid-webhooks dependency from requirements.txt

Molier committed:
chore: ran the gen_requirements

Molier committed:
chore: ran the gen_requirements

Molier committed:
feat: enhance entity suggestion logic and add initial message to EID when mapping is added giving last known state

Molier committed:
feat: enhance entity suggestion logic and add initial message to EID when mapping is added giving last known state

Molier committed:
fix: add diagnostics, make sure 95code cov , hit plat quality

Molier committed:
fix: add diagnostics, make sure 95code cov , hit plat quality

Molier commented:
updated the integration, updated the docs PR as well

NoRi2909 commented on homeassistant/components/energyid/strings.json:
```diff
@@ -0,0 +1,151 @@
+    "step": {
+      "user": {
+        "title": "Connect to EnergyID (Step 1 of 3)",
+        "description": "Enter your EnergyID Webhook Provisioning Key and Secret. Find these in your EnergyID portal under Device Provisioning or Webhook settings. More info: [EnergyID Docs]({docs_url})",
+        "data": {
+          "provisioning_key": "Provisioning Key",
+          "provisioning_secret": "Provisioning Secret"
```
In Home Assistant sentence-casing is mandatory, so there is a lot to fix in the current strings. E.g. on these lines:

        "description": "Enter your EnergyID webhook provisioning key and secret. Find these in your EnergyID portal under Device Provisioning or Webhook settings. More info: [EnergyID Docs]({docs_url})",
        "data": {
          "provisioning_key": "Provisioning key",
          "provisioning_secret": "Provisioning secret"

NoRi2909 commented on homeassistant/components/energyid/strings.json:
```diff
@@ -0,0 +1,151 @@
+        }
+      },
+      "auth_and_claim": {
+        "title": "Claim Device in EnergyID (Step 2 of 3)",
```
"title": "Claim device in EnergyID (Step 2 of 3)",

NoRi2909 commented on homeassistant/components/energyid/strings.json:
```diff
@@ -0,0 +1,151 @@
+        "data": {}
+      },
+      "finalize": {
+        "title": "Finalize Setup (Step 3 of 3)",
```
"title": "Finalize setup (Step 3 of 3)",

NoRi2909 commented on homeassistant/components/energyid/strings.json:
```diff
@@ -0,0 +1,151 @@
+        "title": "Finalize Setup (Step 3 of 3)",
+        "description": "Successfully connected to EnergyID!\n\nPlease confirm or set the name this Home Assistant instance should use when communicating with EnergyID. This name will appear in your EnergyID webhook device list, helping you identify this connection.",
+        "data": {
+          "device_name": "Device Name (for EnergyID Webhook)"
```
"device_name": "Device name (for EnergyID webhook)"

NoRi2909 commented on homeassistant/components/energyid/strings.json:
```diff
@@ -0,0 +1,151 @@
+        }
+      },
+      "reconfigure": {
+        "title": "Reconfigure EnergyID Connection",
```
"title": "Reconfigure EnergyID connection",

NoRi2909 commented on homeassistant/components/energyid/strings.json:
```diff
@@ -0,0 +1,151 @@
+      "init": {
+        "title": "Manage EnergyID Mappings",
+        "data": {
+          "next_step": "Select Action"
```
"title": "Manage EnergyID mappings",
        "data": {
          "next_step": "Select action"

NoRi2909 replied:
Etc. on the following lines.

Molier committed:
chore: fix sentence-casing for strings

Molier committed:
chore: fix sentence-casing for strings

NoRi2909 reviewed (commented):
The strings look OK to me now. I cannot approve the full PR as I'm just a translator.

JrtPec reviewed (approved):
Reviewed the code and did a functional test. Was able to configure the integration and linked a power sensor in HA. Values were uploaded every minute to my EnergyID-account. Works great!

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+            return "cannot_retrieve_claim_info"
+        return "needs_claim"
+
+    async def async_step_reconfigure(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a reconfiguration flow initialized by the user."""
+        errors: dict[str, str] = {}
+
+        if not self._config_entry_being_reconfigured:
+            entry_id = self.context.get("entry_id")
+            if not entry_id:
+                _LOGGER.error("Reconfigure flow started without entry_id in context")
+                return self.async_abort(reason="unknown_error")
+            config_entry = self.hass.config_entries.async_get_entry(entry_id)
+            if not config_entry:
+                _LOGGER.error("Config entry %s not found for reconfigure", entry_id)
```
remove that for now, and leave for follow up

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+            description_placeholders=placeholders_for_finalize,
+        )
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(
+        config_entry: EnergyIDConfigEntry,
+    ) -> EnergyIDSubentryFlowHandler:
+        """Return the options flow handler for the EnergyID integration."""
+        return EnergyIDSubentryFlowHandler()
```
also follow up

Molier replied:
Hi thanks for the review! I'm a bit unsure on some points, you want me to defer some parts to next PRs to make the scope of the initial integration smaller if i understand it? For the diagnostics and the reconfigure flow fe.? But the subconfig flow is quite integral to the integration so im not sure what to do with this.
The main config flow is to create the connection and the device that will interact with our API. And afterwards you use the subconfig flow to add mappings for data to be sent to EID. So without it its not really a functional integration?

I saw this similar setup in other integrations as well: create the initial integration first and then create "entities" for it? but here those are just behind-the-scenes mappings. So not sure how to proceed on this. 

Also just to confirm the reasoning for deferring parts to later PR since they work now as well and made it so the integration could start its live with the plat certification ;)

Molier replied:
I believe I understand your points now:

1.  For managing the individual sensor mappings (where each mapping links an HA entity to an EnergyID key), the **Subentry Flow** pattern (using `async_get_supported_subentry_types` in `EnergyIDConfigFlow` and a separate handler class inheriting from `ConfigSubentryFlow`) is indeed the more appropriate architectural pattern than the `OptionsFlow` I'm currently using via `async_get_options_flow`.

And to clear it up:

2.  To reduce the scope of *this initial PR*, you'd like me to:
    *   Remove the current mapping configuration UI and its underlying logic entirely for now (i.e., remove `async_get_options_flow` and the `EnergyIDSubentryFlowHandler`).
    *   This means the first version of the integration merged will establish the connection to EnergyID and create the device, but will **not yet provide a UI for users to add, edit, or manage any sensor mappings**, and therefore won't be sending sensor-specific data based on user configuration yet.
    *   Defer the reconfigure flow and diagnostics as previously discussed.

3.  Then, in a **follow-up PR**, I will re-introduce the sensor mapping configuration feature, but this time implemented correctly using the **Subentry Flow** pattern.

Does this accurately capture your intended path forward for this integration?

Or can i add the sensor mapping configuration in this PR already as well?

> Yes, the idea is to reduce the initial scope, there is no such thing as an integration starting directly in platinum 😉
> 
> But this is the method for the Options flow (meaning extra config), for subentries the correct one should be `async_get_supported_subentry_types` [developers.home-assistant.io/docs/config_entries_config_flow_handler#subentry-flows](https://developers.home-assistant.io/docs/config_entries_config_flow_handler#subentry-flows)

Molier replied:
@zweckj for some easier more direct communication I am also in the HASS discord, If I'm still misunderstanding some core concepts, we can always talk there. Thanks for reviewing as well :)

zweckj replied:
> Or can i add the sensor mapping configuration in this PR already as well?

Ideally we make the integration as small as possible while still staying viable, so MVP. If I get your integration correctly without the sensor mapping it would basically do nothing, so adding that would be fine for me - it should just use the correct handlers.

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+class EnergyIDConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle the configuration flow for the EnergyID integration."""
+
+    VERSION = 1
```
```suggestion
```
default

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+
+def _generate_energyid_device_id_for_webhook() -> str:
+    """Generate a unique device ID for this Home Assistant instance to use with EnergyID webhook."""
+    return f"{ENERGYID_DEVICE_ID_FOR_WEBHOOK_PREFIX}{secrets.token_hex(4)}"
```
probably better to use the `instance_id` here, so it is guaranteed unique per instance

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+        _LOGGER.debug("User step input: %s", user_input)
+
+        if user_input is not None:
+            self._flow_data["provisioning_key"] = user_input[CONF_PROVISIONING_KEY]
+            self._flow_data["provisioning_secret"] = user_input[
+                CONF_PROVISIONING_SECRET
+            ]
```
```suggestion
            self._flow_data |= user_input
```

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+                    }
+                )
+                return await self.async_step_finalize()
+            if auth_status == "needs_claim":
+                if not self._flow_data.get("claim_info"):
+                    _LOGGER.error("Claim info is missing despite 'needs_claim' status")
+                    return self.async_abort(reason="internal_error_no_claim_info")
```
shouldn't that raise an exception we can catch here?

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+        try:
+            is_claimed = await client.authenticate()
+        except ClientError:
+            _LOGGER.warning(
+                "Connection error during EnergyID authentication", exc_info=True
+            )
```
don't log that
```suggestion
```

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+
+    async def _perform_auth_and_get_details(self) -> str | None:
+        """Authenticate with EnergyID and retrieve device details."""
+        if (
+            not self._flow_data["provisioning_key"]
+            or not self._flow_data["provisioning_secret"]
+        ):
+            _LOGGER.error("Missing credentials for authentication")
+            return "missing_credentials"
```
both are required can not happen

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+            self._flow_data["webhook_device_name"],
+        )
+
+        session = async_get_clientsession(self.hass)
+        client = WebhookClient(
+            provisioning_key=self._flow_data["provisioning_key"],
+            provisioning_secret=self._flow_data["provisioning_secret"],
+            device_id=self._flow_data["webhook_device_id"],
+            device_name=self._flow_data["webhook_device_name"],
+            session=session,
```
```suggestion
        client = WebhookClient(
            provisioning_key=self._flow_data["provisioning_key"],
            provisioning_secret=self._flow_data["provisioning_secret"],
            device_id=self._flow_data["webhook_device_id"],
            device_name=self._flow_data["webhook_device_name"],
            session=async_get_clientsession(self.hass),
```
single use var

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+                client.recordName,
+            )
+            if not self._flow_data["record_number"]:
+                _LOGGER.error("Claimed, but no record number received from EnergyID")
```
```suggestion
```
we also already show this in the UI

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+            self._flow_data["record_number"] = client.recordNumber
+            self._flow_data["record_name"] = client.recordName
+            self._flow_data["claim_info"] = None
+            _LOGGER.info(
```
```suggestion
            _LOGGER.debug(
```
info is reserved for the system

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+        self._flow_data["claim_info"] = claim_details_dict
+        _LOGGER.info("Device needs to be claimed. Claim info: %s", claim_details_dict)
+        if not claim_details_dict or not claim_details_dict.get("claim_code"):
+            _LOGGER.error(
+                "Failed to retrieve valid claim code. Info: %s", claim_details_dict
+            )
```
```suggestion
```
you'll also show what's wrong in the UI

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+
+        claim_details_dict = client.get_claim_info()
+        self._flow_data["claim_info"] = claim_details_dict
+        _LOGGER.info("Device needs to be claimed. Claim info: %s", claim_details_dict)
```
```suggestion
        _LOGGER.debug("Device needs to be claimed. Claim info: %s", claim_details_dict)
```

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+            _LOGGER.debug("Authentication status after claim attempt: %s", auth_status)
+            if auth_status is None:
+                if not self._flow_data.get("record_number"):
+                    _LOGGER.error("Claim successful but record number is missing")
```
```suggestion
```

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+            else:
+                errors["base"] = auth_status
+
+        placeholders_for_form = {
+            "claim_url": "N/A",
+            "claim_code": "N/A",
+            "valid_until": "N/A",
+        }
+        current_claim_info = self._flow_data.get("claim_info")
+        if isinstance(current_claim_info, dict):
+            placeholders_for_form.update(
+                {
+                    "claim_url": current_claim_info.get("claim_url", "N/A"),
+                    "claim_code": current_claim_info.get("claim_code", "N/A"),
+                    "valid_until": current_claim_info.get("valid_until", "N/A"),
+                }
```
what are you trying to achieve here?

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+            "webhook_device_id",
+            "record_number",
+        ]
+        if not all(self._flow_data.get(k) for k in required_keys):
+            _LOGGER.error("Incomplete flow data for finalize: %s", self._flow_data)
+            return self.async_abort(reason="internal_flow_data_missing")
```
how can it happen that we are missing something here?

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+            )
+
+        suggested_name = self._flow_data.get("record_name")
+        if not suggested_name or str(suggested_name).lower() == "none":
+            suggested_name = self._flow_data["webhook_device_name"]
```
```suggestion
        if not suggested_name:
            suggested_name = self._flow_data["webhook_device_name"]
```

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,365 @@
+        if not suggested_name or str(suggested_name).lower() == "none":
+            suggested_name = self._flow_data["webhook_device_name"]
+
+        ha_title_value = self._flow_data.get("record_name") or "your EnergyID site"
+        placeholders_for_finalize = {"ha_entry_title_to_be": str(ha_title_value)}
+
+        return self.async_show_form(
+            step_id="finalize",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_DEVICE_NAME, default=str(suggested_name)): str,
+                }
```
why are you using different names here, what's each one of those?

Molier committed:
refactor(config_flow): Use subentry flow for sensor mappings, also addressed many comments by reviewer

Molier committed:
refactor(config_flow): Use subentry flow for sensor mappings, also addressed many comments by reviewer

Molier commented:
Hi @zweckj,

Thank you for your detailed guidance on using the correct handlers. I've now completed the refactor to use the `Subentry Flow` pattern for managing the sensor mappings.

During the implementation, I ran into a challenging conflict between the `mypy` static type checker and the runtime behavior for this "lightweight" subentry pattern (where the mappings are stored in the parent's `options`).

*   Following the `mypy`-compliant examples (`kitchen_sink`, `mqtt`) of returning the handler class directly (`return {"key": HandlerClass}`) resulted in a runtime `AttributeError`, as `self.config_entry` was not available when the flow started.
*   To fix the runtime error, a `lambda` factory (`lambda: Handler(config_entry)`) is necessary to correctly pass the `config_entry` during instantiation. However, this causes `mypy` to fail due to an `incompatible return type` error.

The final solution I've implemented uses the `lambda` factory to ensure the code works correctly at runtime, and adds a `# type: ignore[]` comment. This approach satisfies all the pre-commit checks and appears to be the standard way to handle this specific architectural inconsistency in Home Assistant.

But I still feel I`m not quite grasping the subentry flow

I have also addressed your other feedback:
*   The `reconfigure` flow has been removed for now.
*   Logging has been updated to use `debug` and remove redundancy.
*   The unique ID now uses the HA `instance_id`.
*   Other code style suggestions have been applied.

This should now be ready for another look when you have a moment. Thanks again for your help and patience!

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,240 @@
+            "webhook_device_name"
+        )
+        placeholders = {
+            "ha_entry_title_to_be": str(
```
I don't think you have that placeholder

Molier committed:
refactor(energyid): streamline config flow and remove diagnostics module,

Molier committed:
refactor(energyid): streamline config flow and remove diagnostics module,

Molier committed:
fix: remove non_functional redundant import of EnergyIDConfigEntry

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,167 @@
+class EnergyIDConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle the configuration flow for the EnergyID integration."""
+
+    VERSION = 1
```
```suggestion
```

zweckj commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,167 @@
+        )
+        try:
+            is_claimed = await client.authenticate()
+            _LOGGER.debug("Authentication successful, claimed: %s", is_claimed)
```
move that out of try

Molier committed:
refactor(const): reorganize constants and remove unused entries

Molier committed:
refactor(const): add missing constants for listeners and mappings

Molier committed:
refactor(energyid): simplify mapping validation by deriving energyid_key from ha_entity_id

Molier committed:
refactor(config_flow): remove unused VERSION attribute from EnergyIDConfigFlow

Molier commented on homeassistant/components/energyid/__init__.py:
```diff
@@ -0,0 +1,427 @@
+
+    async def _close_entry_client(*_: Any) -> None:
+        _LOGGER.debug("Closing EnergyID client for %s", entry.runtime_data.device_name)
+        await entry.runtime_data.close()
```
its a network operation? ill put in a try except then.

Molier commented on homeassistant/components/energyid/sensor.py:
```diff
@@ -0,0 +1,106 @@
+    @property
+    def native_value(self) -> int:
+        """Return the number of active sensor mappings."""
+        return len(self._entry.subentries)
```
makes sense, removed, ty

Molier replied:
@zweckj so you would remove the entire sensor? since we are keeping all the mappings nicely in the more modern subentries? makes sense since the sensor was there originally solely since there was not good managemnt of the mappings in the beginning.

zweckj replied:
I think that sounds reasonable

Molier commented on homeassistant/components/energyid/__init__.py:
```diff
@@ -0,0 +1,427 @@
+        timestamp = timestamp.astimezone(dt.UTC)
+
+    # Create async task to send data to EnergyID
+    hass.async_create_task(
+        client.update_sensor(energyid_key, value, timestamp),
+        name=f"energyid_update_{entity_id}",
+    )
```
i have already implemented that in the webhook client, forgot to use it! thanks haha

Molier committed:
refactor(energyid): streamline runtime data handling and improve state management, added in queuing, fixed race conditions with sending updates, removed extra_state_attributes from the sensor

Molier committed:
fix: small fix

Molier committed:
refactor: remove unused sensor platform and related code

Molier committed:
chore: undid wrong package constraint

Molier commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,167 @@
+            ),
+            errors=errors,
+            description_placeholders={
+                "docs_url": "https://help.energyid.eu/nl/integraties/home-assistant/"
```
i made it more user friendly and now link to where they can get their key and secret. but will further only ref HA docs when needed. thanks

emontnemery commented on homeassistant/components/energyid/energyid_sensor_mapping_flow.py:
```diff
@@ -0,0 +1,156 @@
+            )
+        )
+        current_state = hass.states.get(entity_entry.entity_id)
+        if current_state and current_state.state not in (
+            STATE_UNKNOWN,
+            STATE_UNAVAILABLE,
+        ):
+            try:
+                float(current_state.state)
+            except (ValueError, TypeError):
+                _LOGGER.debug(
+                    "Entity %s state cannot be converted to float",
+                    entity_entry.entity_id,
+                )
+                continue
+            suitable_entities.append(entity_entry.entity_id)
+            _LOGGER.debug(
+                "Added entity %s to suitable entities", entity_entry.entity_id
+            )
+        elif (
+            is_likely_numeric
+            and current_state
+            and current_state.state != STATE_UNAVAILABLE
+        ):
+            suitable_entities.append(entity_entry.entity_id)
+            _LOGGER.debug(
+                "Added likely numeric entity %s to suitable entities",
+                entity_entry.entity_id,
+            )
```
I don't think these checks are needed, the sensor base class raises if sensors have non numeric values since 2 years: https://github.com/home-assistant/core/blob/815b26dcac4668be1bb32eb65730e5718e291bbd/homeassistant/components/sensor/__init__.py#L682-L687

emontnemery reviewed (commented):
A few additional comments.
I've not reviewed the tests

emontnemery commented:
@Molier There are unresolved comments from previous reviews, please address them.
You can message me on Discord if you want to discuss review comments or need some help 👍

Molier committed:
Merge remote-tracking branch 'upstream/dev' into energyid-webhooks

Molier committed:
Address emontnemery code review feedback

- Remove unused constants and signal dispatcher code
- Move cleanup logic to async_unload_entry
- Use modern type annotations and add future imports
- Simplify numeric validation logic
- Remove meaningless comments
- Move DEFAULT_UPLOAD_INTERVAL to __init__.py

Molier committed:
chore: only used strings remain

Molier committed:
fix: update documentation URLs in EnergyID config flow and strings

Molier committed:
feat: added missing string back

Molier committed:
fix: update title format for EnergyID connection in sensor mapping flow

Molier committed:
fix: enhance error handling and cleanup in async_unload_entry for EnergyID integration

Molier committed:
fix: remove unnecessary data field in auth_and_claim description for EnergyID connection

Molier committed:
feat: created updated tests for new version of integration

Molier committed:
Merge branch 'dev' into energyid-webhooks

Molier committed:
fix: remove data_schema from snapshots in config flow tests

emontnemery reviewed (commented):
Some more comments. I've not reviewed the tests.

Molier commented on homeassistant/components/energyid/__init__.py:
```diff
@@ -0,0 +1,268 @@
+    ):
+        return
+
+    entry = hass.config_entries.async_get_entry(entry_id)
+    if not entry or not hasattr(entry, "runtime_data"):
+        _LOGGER.debug(
+            "State change for %s ignored: entry %s not ready or unloading",
+            entity_id,
+            entry_id,
+        )
+        return
```
yes i was having some race conditions during testing sometimes, ill clarify

emontnemery replied:
The debug print is removed (up to you), but there's no clarifying comment

Molier committed:
fix: enhance error handling during client authentication and update state listener management

Molier commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,165 @@
+            },
+        )
+
+    async def async_step_auth_and_claim(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the step for device claiming if needed."""
+        _LOGGER.debug("Starting auth and claim step with input: %s", user_input)
+        if user_input is not None:
+            auth_status = await self._perform_auth_and_get_details()
+
+            if auth_status is None:
+                await self.async_set_unique_id(self._flow_data["record_number"])
+                self._abort_if_unique_id_configured()
+                _LOGGER.debug(
+                    "Creating entry with title: %s", self._flow_data["record_name"]
+                )
+                return self.async_create_entry(
+                    title=self._flow_data["record_name"], data=self._flow_data
+                )
+
+            _LOGGER.debug(
+                "Claim failed or timed out, errors: %s",
+                {"base": "claim_failed_or_timed_out"},
+            )
+            return self.async_show_form(
+                step_id="auth_and_claim",
+                description_placeholders=self._flow_data.get("claim_info", {}),
+                errors={"base": "claim_failed_or_timed_out"},
+            )
+
+        return self.async_show_form(
+            step_id="auth_and_claim",
+            description_placeholders=self._flow_data.get("claim_info", {}),
+        )
```
i tested some stuff out and used combination of external step with polling, that does seem to add some value , removes 1 button press ;)

Molier committed:
feat: implement polling mechanism for device claiming in EnergyID integration

Molier committed:
fix: update entity ID handling to use UUID for EnergyID integration

Molier committed:
Refactor EnergyID integration tests for improved clarity and coverage

- Updated test descriptions for better readability.
- Enhanced mock configuration entries with additional attributes.
- Consolidated setup logic for energyid integration tests.
- Improved error handling tests for various scenarios including entity not found and already mapped entities.
- Added comprehensive tests for async setup, unload, and state change handling.
- Ensured proper handling of non-numeric states and unavailable states in the state change handler.
- Streamlined the creation of mock entities and subentries for testing.

Molier committed:
test: ensured 99 code cov tests

Molier committed:
refactor: log-once for unavailability , brought qualityscale up to date and updated tests to match new pattern. Also fixed the subentry mapping to have it work. added reauth flow

Molier committed:
fix: update mock listener assertion in async unload entry test

emontnemery commented on homeassistant/components/energyid/__init__.py:
```diff
@@ -0,0 +1,268 @@
+        """Authenticate the client and handle errors appropriately."""
+        try:
+            is_claimed = await client.authenticate()
+        except Exception as err:
```
Don't we want to log unexpected exceptions? (It's OK i we don't want do do that for some reason, I'm just checking so it was not forgotten)

emontnemery commented on homeassistant/components/energyid/__init__.py:
```diff
@@ -0,0 +1,268 @@
+    await async_update_listeners(hass, entry)
+
+
+async def async_update_listeners(
+    hass: HomeAssistant, entry: EnergyIDConfigEntry
+) -> None:
```
Has this been addressed?

Molier committed:
fix: clarify return for race cond

Molier committed:
refactor: enhance entity listener management to track uuids better and be resistent to entity name changes

Molier commented on homeassistant/components/energyid/__init__.py:
```diff
@@ -0,0 +1,268 @@
+
+    if (
+        not entity_id
+        or not new_state
```
When new_state is None, the entity was removed from the state machine. We don't need special handling - just skip processing since there's no value to send to EnergyID.

Molier commented on homeassistant/components/energyid/__init__.py:
```diff
@@ -0,0 +1,268 @@
+
+    runtime_data = entry.runtime_data
+    if not (energyid_key := runtime_data.mappings.get(entity_id)):
+        return
```
Added comment. This guards against the entity no longer being mapped (e.g., user just changed options to unmap it, but we're still processing queued events from before the change).

Molier committed:
refactor(energyid): address code review feedback for __init__.py

- Replace generic listeners dict with typed dataclass fields
- Remove unused constants (LISTENER_KEY_*)
- Make update_listeners synchronous (@callback) as it doesn't await
- Fix registry listener to properly recreate when tracked entities change
- Change boot-time warning to debug (expected behavior)
- Add clarifying comments for race condition guards in state handler

emontnemery reviewed (commented):
I'm pretty happy with the listeners now, some minor remarks though.

Molier committed:
fix:Address review feedback: refine energyid logging, comments, and event handling

Molier committed:
remove uv.lock

Molier committed:
refactor: streamline entity registry update handling and improve logging

NoRi2909 commented on homeassistant/components/energyid/strings.json:
```diff
@@ -0,0 +1,66 @@
+      },
+      "auth_and_claim": {
+        "title": "Claim device in EnergyID",
+        "description": "This Home Assistant connection needs to be claimed in your EnergyID portal before it can send data.\n\n1. Go to: {claim_url}\n2. Enter code: **{claim_code}**\n3. (Code expires: {valid_until})\n\nAfter successfully claiming the device in EnergyID, click **Submit** below to continue."
```
We try to avoid "click" for touch-devices:

```suggestion
        "description": "This Home Assistant connection needs to be claimed in your EnergyID portal before it can send data.\n\n1. Go to: {claim_url}\n2. Enter code: **{claim_code}**\n3. (Code expires: {valid_until})\n\nAfter successfully claiming the device in EnergyID, select **Submit** below to continue."
```

NoRi2909 commented on homeassistant/components/energyid/strings.json:
```diff
@@ -0,0 +1,66 @@
+        "title": "Reauthenticate EnergyID",
+        "description": "Please re-enter your EnergyID provisioning key and secret to restore the connection.\n\nMore info: {docs_url}",
+        "data": {
+          "provisioning_key": "Provisioning key",
+          "provisioning_secret": "Provisioning secret"
+        },
+        "data_description": {
+          "provisioning_key": "Your unique key for provisioning.",
+          "provisioning_secret": "Your secret associated with the provisioning key."
```
These should be deduplicated by using references instead:

```suggestion
          "provisioning_key": "[%key:component::energyid::config::step::user::data::provisioning_key%]",
          "provisioning_secret": "[%key:component::energyid::config::step::user::data::provisioning_secret%"
        },
        "data_description": {
          "provisioning_key": "[%key:component::energyid::config::step::user::data_description::provisioning_key%]",
          "provisioning_secret": "[%key:component::energyid::config::step::user::data_description::provisioning_secret%"``

NoRi2909 reviewed (commented):
Some more small improvements in the user-facing strings.

JrtPec committed:
fix: add error handling for 401 invalid auth

JrtPec commented on homeassistant/components/energyid/config_flow.py:
```diff
@@ -0,0 +1,260 @@
+        try:
+            is_claimed = await client.authenticate()
+            _LOGGER.debug("Authentication successful, claimed: %s", is_claimed)
+        except ClientError as err:
+            _LOGGER.error(
+                "Failed to connect to EnergyID during authentication: %s", err
+            )
+            return "cannot_connect"
```
Done, added a check for ClientResponseError with status 401 with a separate error message.

JrtPec committed:
Update homeassistant/components/energyid/strings.json

"select" instead of "click"

Co-authored-by: Norbert Rittel <norbert@rittel.de>

JrtPec committed:
Update homeassistant/components/energyid/strings.json

Replace provisioning key and secret strings with references

Co-authored-by: Norbert Rittel <norbert@rittel.de>

Molier committed:
Merge remote-tracking branch 'upstream/dev' into energyid-webhooks

Molier committed:
feat: Improve EnergyID sync and enhance test coverage

Refactor async_setup_entry to use async_track_time_interval for periodic sensor synchronization instead of a background task, ensuring proper cleanup on unload. Add comprehensive tests for EnergyID integration including config flow error handling (e.g., ClientResponseError for 401 and other codes), reauthentication scenarios with device claiming, supported subentry types, sensor mapping flow with UUID error cases, and integration setup/teardown with state management and entity updates. This improves reliability and robustness of the integration while expanding test coverage to prevent regressions.

emontnemery reviewed (changes requested):
Some additional comments.
There are unresolved comments from previous reviews.

Molier commented on homeassistant/components/energyid/quality_scale.yaml:
```diff
@@ -0,0 +1,137 @@
+  # Gold
+  devices:
+    status: done
+    comment: A device entry is created to represent the connection to the EnergyID service.
```
after reading more into the device definition,https://developers.home-assistant.io/docs/device_registry_index/ i see that what i thought was 1 'device' is just the service entry. but so thats not an actual 'device' then? from the docs:"A device in Home Assistant represents either a physical device that has its own control unit, or a service". So i thought it fell under the service category.

emontnemery replied:
You should collect entities tied to a config entry to a device representing that account. To do that, set [device_info on the entities](https://developers.home-assistant.io/docs/device_registry_index)

We don't require that, so it's up to you if you want to add it now or in a follow-up.

Molier committed:
fix: removed legacy result from test

Molier committed:
refactor: consolidate error handling tests for ClientResponseError

Molier committed:
feat: enhance config flow with unique device ID to allow duplicate entries for different eid connections

Molier committed:
chore: removed unused fixture

Molier committed:
refactor: simplify tests for suggested entities and remove redundant cases, addressing all test sensor mapping flow review remarks

Molier committed:
refactor: update flow logic concerning unique ID needed for eid and unique ID needed for HASS, allowing for multiple connections to multiple Records in EID.

Molier committed:
refactor: remove redundant exception handling for ConfigEntryNotReady in async_setup_entry

Molier committed:
Merge branch 'dev' into energyid-webhooks

Molier committed:
refactor: add sensor mapping hint after successful adding of service entry

emontnemery commented on homeassistant/components/energyid/quality_scale.yaml:
```diff
@@ -0,0 +1,137 @@
+  # Gold
+  devices:
+    status: exempt
+    comment: The integration represents a service connection as a single device entry, created automatically by Home Assistant when the config entry is set up. This service device groups the mapped sensor entities but does not create individual device entries for each sensor as the sensors belong to their original integrations.
```
I think this is clearer.
```suggestion
    comment: The integration does not create any entities, nor does it create devices.
```

Molier committed:
Update homeassistant/components/energyid/quality_scale.yaml

Co-authored-by: Erik Montnemery <erik@montnemery.com>

Molier committed:
Update homeassistant/components/energyid/__init__.py

clarifying comment

Co-authored-by: Erik Montnemery <erik@montnemery.com>

Molier committed:
refactor: change error logging to debug level for EnergyID authentication

Molier committed:
Merge branch 'dev' into energyid-webhooks

Molier committed:
docs: add clarification comments for device_id usage in webhook system

emontnemery reviewed (dismissed):
LGTM, thanks @Molier 👍

Molier committed:
fix: change some strings and test stuff for updated code

Molier commented:
tests pass locally, i cant rerun CI i think?

emontnemery commented:
There are some uncovered lines in `homeassistant/components/energyid/config_flow.py` and `homeassistant/components/energyid/energyid_sensor_mapping_flow_py`, please add tests covering them.
Considering the nature of this integration, I also think we should have full test coverage of `homeassistant/components/energyid/__init__.py`

To check test coverage locally, run `pytest --cov=homeassistant.components.energyid --cov-report html tests/components/energyid`

Molier committed:
test: enhance EnergyID integration tests with error handling and validation

- Added tests for polling behavior on authentication errors and connection issues in `test_config_flow.py`.
- Implemented validation tests for entity mapping inputs in `test_energyid_sensor_mapping_flow.py`.
- Enhanced initialization tests in `test_init.py` to cover timeout scenarios, periodic sync errors, and entry unloading with subentries.
- Improved error handling in various test cases to ensure robustness against unexpected exceptions and state changes.

Molier committed:
test: add error handling for initial state conversion and entry unload scenarios. 100 code cov

Molier commented:
> There are some uncovered lines in `homeassistant/components/energyid/config_flow.py` and `homeassistant/components/energyid/energyid_sensor_mapping_flow_py`, please add tests covering them. Considering the nature of this integration, I also think we should have full test coverage of `homeassistant/components/energyid/__init__.py`
> 
> To check test coverage locally, run `pytest --cov=homeassistant.components.energyid --cov-report html tests/components/energyid`

added tests for full coverage and added some more advanced test for __init__

Molier committed:
chore: remove remnants of dev comments

Molier committed:
Update strings.json

Molier committed:
feat: update integration name references in EnergyID config flow and strings

emontnemery commented on tests/components/energyid/test_config_flow.py:
```diff
@@ -0,0 +1,775 @@
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": config_entries.SOURCE_USER}
+        )
```
Check the result is as expected, don't just assign and discard. Same for the other tests

emontnemery commented on tests/components/energyid/test_config_flow.py:
```diff
@@ -0,0 +1,775 @@
+        "homeassistant.components.energyid.config_flow.WebhookClient",
+        return_value=mock_client,
+    ):
+        result = await hass.config_entries.flow.async_init(
```
Check it

emontnemery commented on tests/components/energyid/test_init.py:
```diff
@@ -0,0 +1,1328 @@
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+    await hass.async_block_till_done()
+
+    assert mock_config_entry.state is ConfigEntryState.SETUP_ERROR
```
These cases should check a reauth flow is started

Molier committed:
refactor: reorganize constants and improve logging in EnergyID config flow

Molier committed:
Merge branch 'dev' into energyid-webhooks

Molier committed:
fix: improve error handling during EnergyID authentication process

Molier committed:
refactor: enhance comment on polling timeout to clarify user notification reasoning

Molier committed:
refactor: update reauthentication success message and add entry type for sensor mapping

Molier commented on tests/components/energyid/test_energyid_sensor_mapping_flow.py:
```diff
@@ -0,0 +1,397 @@
+    hass.states.async_set(entity_entry.entity_id, "1500")
+
+    result = _get_suggested_entities(hass)
+    assert entity_entry.entity_id in result
```
yea you're right, both were actually testing the same `original_device_class` branch but with different device classes, remnants from trying desperately to cover all lines. I've consolidated the two tests into a single parametrized test that now covers both:

- Entities with original device class set by integration (ENERGY/POWER)
- Entities with user-overridden device class

Molier committed:
refactor: enhance test for suggested entities with parameterized device class cases

Molier commented on tests/components/energyid/test_config_flow.py:
```diff
@@ -0,0 +1,775 @@
+        )
+        await hass.async_block_till_done()
+
+        assert result2["type"] is FlowResultType.CREATE_ENTRY
```
check added

Molier committed:
fix:address reviewer comments and achieve 100 test cov.
  - Verify reauth flows in error scenarios
  - Add missing assertions and remove redundant tests
  - Cover polling task cancellation edge cases
  - Fix missing translations

Molier committed:
Merge branch 'dev' into energyid-webhooks

Molier committed:
refactor: set POLLING_INTERVAL  for faster more reliable tests and also added 1 more assert as suggested

emontnemery reviewed (approved):
This looks good to me now, thanks @Molier 👍
